sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java

```diff
@@ -0,0 +1,684 @@
+      if (replica != null) {
+        return getEstimatedSizeOfShard(replica);
+      } else {
+        return getEstimatedSizeOfCollection();
+      }
+    }
+
+    private long getEstimatedSizeOfShard(@Nonnull ReplicaInfo replica) throws IOException {
+      try (AuthorizedSolrClient solrClient = spec.getConnectionConfiguration()
+          .createClient(replica.baseUrl())) {
+        CoreAdminRequest req = new CoreAdminRequest();
+        req.setAction(CoreAdminParams.CoreAdminAction.STATUS);
+        req.setIndexInfoNeeded(true);
+        CoreAdminResponse response;
+        try {
+          response = (CoreAdminResponse) solrClient.process(req);
+        } catch (SolrServerException e) {
+          throw new IOException("Can not get core status from " + replica, e);
+        }
+        NamedList<Object> coreStatus = response.getCoreStatus(replica.coreName());
+        NamedList<Object> indexStats = (NamedList<Object>) coreStatus.get("index");
+        return (long) indexStats.get("sizeInBytes");
+      }
+    }
+
+    private long getEstimatedSizeOfCollection() throws IOException {
+      long sizeInBytes = 0;
+      ConnectionConfiguration config = spec.getConnectionConfiguration();
+      try (AuthorizedSolrClient<CloudSolrClient> solrClient = config.createClient()) {
+        DocCollection docCollection = AuthorizedSolrClient.getClusterState(solrClient)
+            .getCollection(config.getCollection());
+        if (docCollection.getSlices().isEmpty()) {
+          return 0;
+        }
+
+        ArrayList<Slice> slices = new ArrayList<>(docCollection.getSlices());
+        Collections.shuffle(slices);
+        ExecutorService executor = Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder()
+                .setThreadFactory(MoreExecutors.platformThreadFactory())
+                .setDaemon(true)
+                .setNameFormat("solrio-size-of-collection-estimation")
+                .build());
+        try {
+          ArrayList<Future<Long>> futures = new ArrayList<>();
+          for (int i = 0; i < 100 && i < slices.size(); i++) {
+            Slice slice = slices.get(i);
+            final Replica replica = slice.getLeader();
+            Future<Long> future = executor.submit(new Callable<Long>() {
+              @Override public Long call() throws Exception {
+                return getEstimatedSizeOfShard(ReplicaInfo.create(replica));
+              }
+            });
+            futures.add(future);
+          }
+          for (Future<Long> future : futures) {
+            try {
+              sizeInBytes += future.get();
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+              throw new IOException(e);
+            } catch (ExecutionException e) {
+              throw new IOException("Can not estimate size of shard", e.getCause());
+            }
+          }
+        } finally {
+          executor.shutdownNow();
+        }
```

Comment:
I'm not a fan this change, because if there are IOException when connect to leader of slice, there must be something happened on Solr cluster. So why should we continue the processing here?